### PR TITLE
Add ComputeElementFlux method to ElasticityIntegrator [elasticity-flux]

### PR DESCRIPTION
**Pull request by barker29**
Add ComputeElementFlux() to ElasticityIntegrator

This can be seen as one step toward implementing a ZZ error estimator for elasticity.

This was implemented as part of the LiDO project, and some of those guys asked me to see if MFEM was interested in including this code. The PR is more a request for comment to see if this is the kind of thing you're interested in than a request to merge exactly this code.

### CODE REVIEW

**Comment by barker29** [comment]
@mlstowell can you suggest another reviewer? I am struggling to find someone appropriate for this.

File: fem/bilininteg.cpp [hunk 1]
```diff
@@ -2254,6 +2254,93 @@ void ElasticityIntegrator::AssembleElementMatrix(
...
+   flux = 0.0;
+   double L, M, E, v;
+   double e11, e22, e33, e12, e13, e23;
+   double s11, s22, s33, s12, s13, s23;
+
+   for (int i = 0; i < fnd; i++)
+   {
+      const IntegrationPoint &ip = ir.IntPoint(i);
+      el.CalcDShape(ip, dshape);
+      DenseMatrix loc_data_mat(u.GetData(), nd, vdim);
```
Comment (mlstowell) [inline fem/bilininteg.cpp]: This is a relatively light weight call but it should be outside the loop.

File: fem/bilininteg.cpp [hunk 2]
```diff
@@ -2254,6 +2254,93 @@ void ElasticityIntegrator::AssembleElementMatrix(
...
+         stress(0,0) = s11; stress(0,1) = s12; stress(0,2) = s13;
+         stress(1,0) = s12; stress(1,1) = s22; stress(1,2) = s23;
+         stress(2,0) = s13; stress(2,1) = s23; stress(2,2) = s33;
+      }
+
+      // the stress values at each integration point in the column
+      // selected by flux_col_ are stored in a vector
+      // in principle, flux_col_ could be selected by the user to give
+      // slightly different error estimatros
+      const int flux_col_ = 0;
```
Comment (mlstowell) [inline fem/bilininteg.cpp]: Perhaps this method should return a DenseMatrix rather than a Vector and flux_col_ could be eliminated.

Comment (mlstowell) [reply fem/bilininteg.cpp]: A more general solution might be to produce an interleaved Vector of some sort which could store the entire matrix of values.

**Comment by mlstowell** [comment]
I can probably track this down in the TopOpt code but for sake of discussion how would this be used?

File: fem/bilininteg.cpp (hunk 2 above)
Comment (barker29) [reply fem/bilininteg.cpp]: The `BilinearFormIntegrator` interface that this implements returns a Vector. Flattening all the values from the stress tensor into a Vector might make sense, but I'm not sure it's very intuitive for what a user will expect. I'm not sure what the best approach here is.

**Comment by barker29** [comment]
In LiDO, this is used as part of a Zienkewicz-Zhu error estimator. The details of that estimator are a bit weird, but since this particular method is implemented in MFEM for DiffusionIntegrator and CurlCurlIntegrator, it may make sense to include it also for ElasticityIntegrator.

**Comment by v-dobrev** [comment]
Thanks for adding the example, Mark. This will be really helpful.

I'll work on adding documentation for the methods `ComputeElementFlux` and `ComputeElementEnergy`.

**Comment by v-dobrev** [comment]
Everyone,

I made some changes and added some new code, including a parallel version of the AMR elasticity example that Mark started. (The new example is numbered 22 since number 21 is used in #584.)

It is probably a good idea for everyone interested to take another look at this PR before merging.

Outstanding tasks:
- [ ] Update `CHANGELOG`
- [ ] Add documentation for the new example in `examples/README.html`
- [ ] Anything else?

File: fem/bilininteg.hpp [hunk 3]
```diff
@@ -69,12 +69,60 @@ class BilinearFormIntegrator : public NonlinearFormIntegrator
                                  const Vector &elfun, DenseMatrix &elmat)
    { AssembleFaceMatrix(el1, el2, Tr, elmat); }
 
+   /** @brief Virtual method required for Zienkiewicz-Zhu type error estimators.
```
Comment (tzanio) [inline fem/bilininteg.hpp]: Awesome documentation work!

**Comment by tzanio** [comment]
A couple of possible images for the documentation

glvis_s11
glvis_s09

**Comment by tzanio** [comment]
Hi-res versions:
glvis_s01
glvis_s06

**Comment by tzanio** [comment]
Companion documentation PR here: https://github.com/mfem/web/pull/17

**Comment by tzanio** [comment]
Re-approved

**Comment by jakubcerveny** [comment]
I added load balancing in ex22p. With the coarse mesh reordering I'm working on, the example should be even cooler on hexes :-)

**Comment by mlstowell** [comment]
I think this is ready to merge.

**Pull request merged**